src/docker-utils.coffee

```diff
@@ -15,6 +17,72 @@ Promise.promisifyAll(docker.getContainer().constructor.prototype)
 
 exports.docker = docker
 
+# Create an array of (repoTag, image_id, created) tuples like the output of `docker images`
+listRepoTagsAsync = ->
+	docker.listImagesAsync()
+	.then (images) ->
+		images = _.sortByOrder(images, 'Created', [ false ])
+		ret = []
+		for image in images
+			for repoTag in image.RepoTags
+				ret.push [ repoTag, image.Id, image.Created ]
+		return ret
+
+# Find either the most recent image of the same app or the image of the supervisor
+findSimilarImage = (repoTag) ->
+	application = repoTag.split('/')[1]
+
+	listRepoTagsAsync()
+	.then (repoTags) ->
+		# Find the most recent image of the same application
+		for repoTag in repoTags
+			otherApplication = repoTag[0].split('/')[1]
+			if otherApplication is application
+				return repoTag
+
+		# Otherwise return the most specific supervisor tag (commit hash)
+		for repoTag in repoTags when /resin\/.*-supervisor.*:[0-9a-f]{6}/.test(repoTag)
+			return repoTag
+
+		return config.supervisorImage
+
+exports.rsyncImageWithProgress = (imgDest, onProgress) ->
+	findSimilarImage(imgDest)
+	.spread (imgSrc) ->
+		rsyncDiff = new Promise (resolve, reject) ->
+			console.log("#{config.deltaHost}/api/v1/delta?src=#{imgSrc}&dest=#{imgDest}", timeout: 5 * 60 * 1000)
+			progress request.get("#{config.deltaHost}/api/v1/delta?src=#{imgSrc}&dest=#{imgDest}", timeout: 5 * 60 * 1000)
+			.on 'progress', onProgress
+			.on 'end', ->
+				onProgress(percent: 100)
+			.on 'response', (res) ->
+				if res.statusCode isnt 200
+					reject()
```

Comment:
Might be good to reject(new Error(res.body)) or something like that to get more info in case of an error?